ipq40xx: Add support for Crisis Innovation Lab MeshPoint.One

robimarko opened the pull request:
MeshPoint.One is Wi-Fi hotspot and smart IoT gateway (based upon Jalapeno module from 8Devices).

MeshPoint.One (https://meshpointone.com) is a unique Wi-Fi hotspot and smart city gateway that can be installed and powered from street lighting (even solar power in the future).
MeshPoint provides up to 27 hours of interrupted Wi-Fi and IoT services from internal battery even when external power is not available.
MeshPoint.One can be used for disaster relief efforts in order to provide instant Wi-Fi coverage that can be easily expanded by just adding more devices that create wide area mesh network.
MeshPoint.One devices have standard Luci UI for management.

Features:
- 1x 1Gpbs WAN
- 1x 1Gbps LAN
- POE input (eth0)
- POE output (eth1)
- Sensor for temperature, humidity and pressure (Bosch BME280)
- current, voltage and power measurement via TI INA230
- Hardware real time clock
- optional power via Li-Ion battery
- micro usb port with usb to serial chip for easy OpenWrt terminal access
- I2C header for connecting additional sensors

Installation:
-------------
Simply flash the sysupgrade image from stock firmware.

Or use the built in Web recovery into bootloader:
Hold Reset button for 5 to 20 seconds or use UART and httpd command.
Web UI will appear on 192.168.2.100 by default.
For web recovery use the factory.ubi image.

Signed-off-by: Robert Marko <robert@meshpoint.me>
Signed-off-by: Damir Samardzic <damir.samardzic@sartura.hr>
Signed-off-by: Damir Franusic <damir.franusic@sartura.hr>
Signed-off-by: Valent Turkovic <valent@meshpoint.me>

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-meshpoint-one.dts:
```diff
@@ -0,0 +1,318 @@
+	leds {
+		compatible = "gpio-leds";
+
+		status: status {
```
Please use
"led_status: status {"
and update aliases.

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-meshpoint-one.dts:
```diff
@@ -0,0 +1,318 @@
+			pins = "gpio52";
+			function = "mdc";
+		};
+		pinconf {
```
Add empty lines between nodes.

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-meshpoint-one.dts:
```diff
@@ -0,0 +1,318 @@
+			drive-strength = <2>;
+			bias-disable;
+		};
+		pin_cs {
```
Add empty line between nodes.

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-meshpoint-one.dts:
```diff
@@ -0,0 +1,318 @@
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition0@0 {
```
Remove the digit before the "@", so you get "partition@0".
Same for the other definitions below, at the end "partition5" is repeated anyway.
Add empty lines between the nodes.

ynezz commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-meshpoint-one.dts:
```diff
@@ -0,0 +1,318 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
```
Looks like it should be handled via DTSI.

```diff
--- target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-meshpoint-one.dts	2019-11-21 15:28:20.414783321 +0100
+++ target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-jalapeno.dts	2019-11-15 07:46:20.300929176 +0100
@@ -1,6 +1,18 @@
-// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
-/* Copyright (c) 2019, CRISIS INNOVATION LAB d.o.o.
- * Author: Robert Marko <robert@meshpoint.me>
+/* Copyright (c) 2015, The Linux Foundation. All rights reserved.
+ * Copyright (c) 2018, Robert Marko <robimarko@gmail.com>
+ *
+ * Permission to use, copy, modify, and/or distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
  */
 
 #include "qcom-ipq4019.dtsi"
@@ -9,15 +21,8 @@
 #include <dt-bindings/soc/qcom,tcsr.h>
 
 / {
-	model = "Crisis Innovation Lab MeshPoint.One";
-	compatible = "cilab,meshpoint-one";
-
-	aliases {
-		led-boot = &status;
-		led-failsafe = &status;
-		led-running = &status;
-		led-upgrade = &status;
-	};
+	model = "8devices Jalapeno";
+	compatible = "8dev,jalapeno";
 
 	soc {
 		rng@22000 {
@@ -89,62 +94,6 @@
 		edma@c080000 {
 			status = "okay";
 		};
-
-		i2c-gpio {
-			compatible = "i2c-gpio";
-			gpios = <&tlmm 0 (GPIO_ACTIVE_HIGH|GPIO_OPEN_DRAIN) /* sda */
-					 &tlmm 4 (GPIO_ACTIVE_HIGH|GPIO_OPEN_DRAIN) /* scl */
-					>;
-			#address-cells = <1>;
-			#size-cells = <0>;
-
-			status = "okay";
-
-			bme280@76 {
-				compatible = "bosch,bme280";
-				reg = <0x76>;
-				status = "okay";
-			};
-
-			pcf2129@51 {
-				compatible = "nxp,pcf2129";
-				reg = <0x51>;
-				status = "okay";
-			};
-
-			ina230@40 {
-				compatible = "ti,ina230";
-				reg = <0x40>;
-				shunt-resistor = <2000>;
-				status = "okay";
-			};
-
-			ina230@44 {
-				compatible = "ti,ina230";
-				reg = <0x44>;
-				shunt-resistor = <2000>;
-				status = "okay";
-			};
-		};
-	};
-
-	keys {
-		compatible = "gpio-keys";
-
-		reset {
-			label = "reset";
-			gpios = <&tlmm 5 GPIO_ACTIVE_LOW>;
-			linux,code = <KEY_RESTART >;
-		};
-	};
-
-	leds {
-		compatible = "gpio-leds";
-
-		status: status {
-			label = "meshpoint-one:blue:status";
-			gpios = <&tlmm 63 GPIO_ACTIVE_HIGH>;
-		};
 	};
 };
 
@@ -166,7 +115,7 @@
 
 	serial_pins: serial_pinmux {
 		mux {
-			pins = "gpio61", "gpio60";
+			pins = "gpio60", "gpio61";
 			function = "blsp_uart0";
 			bias-disable;
 		};
@@ -199,7 +148,7 @@
 	cs-gpios = <&tlmm 54 GPIO_ACTIVE_HIGH>, <&tlmm 59 GPIO_ACTIVE_HIGH>;
 	status = "okay";
 
-	flash@0 {
+	m25p80@0 {
 		reg = <0>;
 		compatible = "jedec,spi-nor";
 		spi-max-frequency = <24000000>;
@@ -270,14 +219,14 @@
 	};
 };
 
-&cryptobam {
+&blsp1_uart1 {
+	pinctrl-0 = <&serial_pins>;
+	pinctrl-names = "default";
 	status = "okay";
 };
 
-&blsp1_uart1 {
+&cryptobam {
 	status = "okay";
-	pinctrl-0 = <&serial_pins>;
-	pinctrl-names = "default";
 };
 
 &gmac0 {
@@ -315,4 +264,3 @@
 &usb2_hs_phy {
 	status = "okay";
 };
-
```

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-meshpoint-one.dts:
```diff
@@ -0,0 +1,318 @@
+		compatible = "spi-nand";
+		reg = <1>;
+		spi-max-frequency = <24000000>;
+		partitions {
```
Add empty lines after "status" and before "partitions".

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-meshpoint-one.dts:
```diff
@@ -0,0 +1,318 @@
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition0@0 {
```
-> "partition@0"

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-meshpoint-one.dts:
```diff
@@ -0,0 +1,318 @@
+&usb2_hs_phy {
+	status = "okay";
+};
+
```
Remove empty line at the end.

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-meshpoint-one.dts:
```diff
@@ -0,0 +1,318 @@
+};
+
+&blsp1_uart1 {
+	status = "okay";
```
add empty line after status.

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-meshpoint-one.dts:
```diff
@@ -0,0 +1,318 @@
+};
+
+&wifi0 {
+	status = "okay";
```
Add empty line after status here and in next node.

adrianschmutzler commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -164,6 +164,18 @@ define Device/avm_fritzrepeater-3000
+	DEVICE_MODEL := MeshPoint.One
+	BLOCKSIZE := 128k
+	PAGESIZE := 2048
+	DEVICE_PACKAGES := kmod-i2c-gpio kmod-iio-bmp280-i2c kmod-hwmon-core kmod-hwmon-ina2xx kmod-rtc-pcf2127
```
hwmon-core should be inherited by kmod-hwmon-ina2xx, so you can remove it.

ynezz commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -164,6 +164,18 @@ define Device/avm_fritzrepeater-3000
 TARGET_DEVICES += avm_fritzrepeater-3000
 
+define Device/cilab_meshpoint-one
+	$(call Device/FitImage)
```
With `$(call Device/8dev_jalapeno)` you can follow DRY principle and reuse 4 common variables, as a bonus you're making it explicit, that this device is based on Jalapeno.

adrianschmutzler replied:
I'm not a big fan of inheriting devices from other Devices. This drove me mad in ar71xx already, and I'm glad this stopped in ath79.
However, with 4 common variables it might qualify for a common definition that can be inherited by both devices (like it's done for Device/netgear_ex61x0v2)...

adrianschmutzler commented:
Hi, unfortunately only formal comments from my side. For the commit message/title:
- Use lower-case "add" in commit title
- wrap your commit message to have less than 75 characters per line
- harmonize your mail address in Signed-off and From

robimarko commented:
@adrianschmutzler @ynezz Your remarks should now be resolved

ynezz commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-jalapeno.dts:
```diff
@@ -1,266 +1,9 @@
- *
- */
+// SPDX-License-Identifier: GPL-2.0-or-later
+// Copyright (c) 2018, Robert Marko <robimarko@gmail.com>
```
`Copyright (c) 2015, The Linux Foundation. All rights reserved.`

ynezz replied:
IANAL, but other copyright statements shouldn't be removed.

robimarko replied:
Well, Linux Foundation has no copyright claim on the file.
I am the sole author and it was most likely leftover in copy/paste.

ynezz commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-jalapeno.dtsi:
```diff
@@ -0,0 +1,269 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
```
Please can you consider adding `SPDX-License-Identifier: GPL-2.0-or-later OR MIT` as in [DTS Checklist](https://openwrt.org/submitting-patches#dts_checklist) ?
> Having MIT as alternative is sometimes preferred by upstream maintainers
and allows sharing that simple code with other projects. We don't really
want multiple DTS versions for the same device.

robimarko replied:
Yes, I added the OR MIT statement but did not push it yet.

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-jalapeno.dts:
```diff
@@ -1,266 +1,9 @@
-&usb2_hs_phy {
-	status = "okay";
-};
+};
```
Missing new line at the end.

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-jalapeno.dtsi:
```diff
@@ -0,0 +1,269 @@
+
+&usb2_hs_phy {
+	status = "okay";
+};
```
Add newline at the end.

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-meshpoint-one.dts:
```diff
@@ -0,0 +1,81 @@
+			gpios = <&tlmm 63 GPIO_ACTIVE_HIGH>;
+		};
+	};
+};
```
Add newline at the end.

adrianschmutzler commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -164,6 +164,15 @@ define Device/avm_fritzrepeater-3000
+	DEVICE_MODEL := MeshPoint.One
+	DEVICE_PACKAGES := kmod-i2c-gpio kmod-iio-bmp280-i2c kmod-hwmon-ina2xx kmod-rtc-pcf2127
+endef
+TARGET_DEVICES += cilab_meshpoint-one
```
As stated earlier, please avoid that and create a common definition for jalapeno board (following the same logic why we have a DTSI and not derive one DTS from the other).

adrianschmutzler commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -164,6 +164,15 @@ define Device/avm_fritzrepeater-3000
 TARGET_DEVICES += avm_fritzrepeater-3000
 
+define Device/cilab_meshpoint-one
+	$(call Device/8dev_jalapeno)
```
As stated earlier, please avoid including another device definition and create a common definition for jalapeno board (following the same logic why we have a DTSI and not derive one DTS from the other).

robimarko replied:
Sure, will do that.

ynezz commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-jalapeno.dts:
```diff
@@ -1,266 +1,9 @@
- * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
- *
- */
+// SPDX-License-Identifier: GPL-2.0-or-later
```
IANAL, but I think, that in order to stay safe it should be `SPDX-License-Identifier: ISC` as this is the license specified in the removed text.

robimarko replied:
Well since I am the author and copyright holder of that DTS, I don't see an issue with changing to GPL2.0+ and MIT combo.

Commit:
ipq40xx: use DTSI for Jalapeno

Lets move common code for Jalapeno into DTSI, this way Jalapeno based boards don't introduce duplicate code.

While at it, lets also fix some style issues and update to current DTS style.

Signed-off-by: Robert Marko <robert@meshpoint.me>

ynezz commented:
LGTM, thanks!

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-meshpoint-one.dts:
```diff
@@ -0,0 +1,81 @@
+		led-upgrade = &led_status;
+	};
+
+	soc {
```
From my experience with other targets, I'd have added a label to soc in the dtsi and used this label here:
&soc {
   i2c-gpio {
       ...
   };
};

I know that both versions work, so this more a comment than a suggestion. Maybe this is also treated differently in ipq40xx ...

robimarko replied:
Well, since upstream and all other DTS code for ipq40xx behaves like this I don't see an advantage in applying another patch to the upstream DTSI and then referencing that node.
A while ago most nodes got labels but soc remained this way.

adrianschmutzler replied:
Fine.

adrianschmutzler commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -60,14 +60,19 @@ define Build/SenaoFW
+	BLOCKSIZE := 128k
+	PAGESIZE := 2048
+endef
+TARGET_DEVICES += 8dev_jalapeno-common
```
TARGET_DEVICES is only required for a real device. Drop this line for the common node here.

robimarko replied:
Thanks, I completely missed this.
Pushed update to fix this.

Commit:
ipq40xx: add support for Crisis Innovation Lab MeshPoint.One

MeshPoint.One is Wi-Fi hotspot and smart IoT gateway (based upon Jalapeno module from 8Devices).

MeshPoint.One (https://meshpointone.com) is a unique Wi-Fi hotspot and smart city gateway
that can be installed and powered from street lighting (even solar power in the future).
MeshPoint provides up to 27 hours of interrupted Wi-Fi and IoT services from internal
battery even when external power is not available.
MeshPoint.One can be used for disaster relief efforts in order to provide instant Wi-Fi
coverage that can be easily expanded by just adding more devices that create wide area mesh network.
MeshPoint.One devices have standard Luci UI for management.

Features:
- 1x 1Gpbs WAN
- 1x 1Gbps LAN
- POE input (eth0)
- POE output (eth1)
- Sensor for temperature, humidity and pressure (Bosch BME280)
- current, voltage and power measurement via TI INA230
- Hardware real time clock
- optional power via Li-Ion battery
- micro usb port with usb to serial chip for easy OpenWrt terminal access
- I2C header for connecting additional sensors

Installation:
-------------
Simply flash the sysupgrade image from stock firmware.

Or use the built in Web recovery into bootloader:
Hold Reset button for 5 to 20 seconds or use UART and httpd command.
Web UI will appear on 192.168.2.100 by default.
For web recovery use the factory.ubi image.

Signed-off-by: Robert Marko <robert@meshpoint.me>
Signed-off-by: Damir Samardzic <damir.samardzic@sartura.hr>
Signed-off-by: Damir Franusic <damir.franusic@sartura.hr>
Signed-off-by: Valent Turkovic <valent@meshpoint.me>

adrianschmutzler commented:
I like that you already improve style when creating the DTSI.

Maybe add a tiny comment about that in the commit message, like
"While at it, fix some style issues." or similar.

robimarko commented:
@adrianschmutzler Good idea.
Pushed updated commit.

adrianschmutzler commented:
LGTM as far as I can judge.

ynezz commented:
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git
